src/NLog/Targets/Target.cs

```diff
@@ -755,6 +790,18 @@ private static bool TryGetCachedValue(Layout layout, LogEventInfo logEvent, out
             return false;
         }
 
+        private static bool TryGetCachedValue<T>(Layout<T> layout, LogEventInfo logEvent, out T value)
+        {
+            if ((!layout.ThreadAgnostic || layout.MutableUnsafe) && logEvent.TryGetCachedLayoutValue(layout, out var value2))
```

Comment:
There is a bug here, where it performs lookup using `Layout<T>`-reference, but the `Layout<T>.PrecalculateBuilder` stores the reference to the inner-layout in the `LogEventInfo._layoutCache`

Reply:
Maybe move this method (and the above method) into the individual Layout-classes. Maybe consider making it a virtual-method and override it in `Layout<T>`-implementation.

Reply:
Actually after further thought then this is a great idea. I have for a long time been worried about storing the raw-value in the LogEventInfo. But actually if `Layout<T>` promise that it will only store values of type `<T>` then it will be a great optimization. Ofcourse right now the current `PrecalculateBuilder´ is not up for the task, but that can be a different PR.